lexer/stateful/stateful.go

```diff
@@ -64,8 +66,151 @@ type compiledRule struct {
+	// fill too much of them (like /./, which would basically force us to allocate all of the runeMap, which
+	// we want to avoid), or because we didn't know their regexps during analysis (backrefs).
+	runeMap     map[int][]bool
+	fallbackMap []bool // a slice the same size as rules, these
+}
+
+func (group *compiledRuleGroup) process() error {
+	var rules = group.rules
+	var nbrules = len(rules)
+	group.runeMap = make(map[int][]bool)
+	group.fallbackMap = make([]bool, nbrules)
+
+	var computed = make([]*computedRuneRange, nbrules)
+	var fallback = group.fallbackMap
+	var runemap = group.runeMap
+
+	// Analyze the patterns to populate runeMap and fallbackMap
+	for i, rule := range rules {
+		if rule.RE == nil || rule.Pattern == "" {
+			fallback[i] = true
+			continue // ignore those
+		}
+		comp, err := computeRuneRanges(rule.Pattern)
+		if err != nil {
+			return err // FIXME better error handling ?
+		}
+		computed[i] = comp
+		if comp.size >= charclassSizeLimit {
+			// this one is too big to have their own map
+			fallback[i] = true
+			continue
+		}
+	}
+
+	// Now process the rules and add them to their respective buckets.
+	for idrule, comp := range computed {
+		if comp == nil || fallback[idrule] {
+			continue
+		}
+		var runes = comp.runes
+		for i, l := 0, len(runes); i < l; i += 2 {
+			for r, end := runes[i], runes[i+1]; r <= end; r++ {
+				var (
+					bools []bool
+					ok    bool
+				)
+				if bools, ok = runemap[r]; !ok {
+					bools = make([]bool, nbrules)
+					runemap[r] = bools
+					// when creating a new bucket, add all the fallback rules onto it if they could potentially
+					for j, isFallback := range fallback {
+						if isFallback {
+							var comp = computed[j]
+							// add all the fallback rules that are either not yet defined (or "")
+							// or those that we know will never match this rune.
+							bools[j] = comp == nil || !runeIsInRange(r, comp)
+						}
+					}
+				}
+
+				bools[idrule] = true // add it onto the bucket.
+			}
+		}
+	}
+
+	return nil
+}
+
+func runeIsInRange(r int, comp *computedRuneRange) bool {
+	if comp == nil {
+		return true // all the unknown rules are added.
+	}
+	var rng = comp.runes
+	for i, l := 0, len(rng); i < l; i += 2 {
+		if r >= rng[i] && r <= rng[i+1] {
+			return true
+		}
+	}
+	return false
+}
+
+// Tries to match to input. Returns nil, nil if no match was found.
+func (group *compiledRuleGroup) tryMatch(l *Lexer) (*compiledRule, []int, error) {
+	// Our goal is to find the indices of the rules to test
+	var (
+		indices []bool
+		ok      bool
+		rules   = group.rules
+		re      *regexp.Regexp
+		err     error
+	)
+
+	// first, get the first rune
+	// FIXME: this is forcing utf-8 as an encoding, maybe do something for other charsets ?
+	r, _ := utf8.DecodeRune(l.data)
+	if r == utf8.RuneError {
+		return nil, nil, fmt.Errorf(`invalid utf-8 input sequence`) // FIXME this should be an error, how do we report it ?
```

Comment:
What does this FIXME mean? Remove?

Reply:
Oh no, it was to have a discussion about encodings. Should we support more ?

Reply:
Ah. No I think it's fine for now.